Copy module: improve tests allowing to use a managed host which isn't the controller host

pilou- opened the pull request:
##### SUMMARY
Copy module integration tests fail when the controller host and the manager host aren't the same host. This pull-request updates the tests in order to be able to use two different hosts.

This issue wasn't spotted by the CI but using the following command:
```
ANSIBLE_REMOTE_USER=root \
ANSIBLE_SSH_ARGS='-o CheckHostIP=no -o VerifyHostKeyDNS=no -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -F /dev/null' \
ANSIBLE_ROLES_PATH=$ANSIBLE_HOME/test/integration/targets \
ansible-playbook -i 'testhost,' -e ansible_host=${REMOTE_IP} -v \
    -e "@$ANSIBLE_HOME/test/integration/integration_config.yml" \
    --tags=test_copy $ANSIBLE_HOME/test/integration/non_destructive.yml 

```

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
copy module: integration tests

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible-playbook 2.4.0 (devel f65ced0438) last updated 2017/06/13 12:19:26 (GMT +200)
```

pilou- committed:
set output_dir_expanded using module result

'path' values are expanded using 'expandvars' too

pilou- committed:
foo.txt is located in 'files' directory

pilou- committed:
Use 'role_path' and 'connection: local' for local paths

'{{ role_path }}/tmp' is used for generated paths

pilou- committed:
Use local connection with local paths

/tmp/ansible-test-abs-link and /tmp/ansible-test-abs-link-dir are
defined by targets/copy/files/subdir/subdir1/ansible-test-abs-link
and targets/copy/files/subdir/subdir1/ansible-test-abs-link-dir links.

pilou- committed:
task names: add a suffix when same name is reused

pilou- committed:
Check that item exists before checking file mode

then error message is more explicit when item doesn't exist

pilou- committed:
Use output_dir_expanded only when necessary

pilou- committed:
Enforce remote_user when root is required

pilou- committed:
Fix remote path

pilou- committed:
Use different local & remote user

this is useful when controller and managed hosts are identical

pilou- commented:
PR updated, ping @abadger.

Tested using two different hosts:
```
ANSIBLE_REMOTE_USER=root \
ANSIBLE_SSH_ARGS='-o CheckHostIP=no -o VerifyHostKeyDNS=no -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -F /dev/null' \
ANSIBLE_ROLES_PATH=$ANSIBLE_HOME/test/integration/targets \
ansible-playbook -i 'testhost,' -e ansible_host=${IP} -v \
    -e "@$ANSIBLE_HOME/test/integration/integration_config.yml" \
    playbook.yml
```
```
$ cat playbook.yml
- hosts: testhost
   gather_facts: True
   roles:
   - { role: copy }
```

abadger commented on test/integration/targets/copy/tasks/main.yml:
```diff
@@ -28,6 +25,9 @@
 
+- set_fact:
+    output_dir_expanded: '{{ file_result_check.path | dirname }}'
+
```
I think this is okay given the way that we currently use it but it doesn't seem right to set a variable that will be used to check correctness of the module from the module itself so we probably shouldn't do it this way (it makes it so there are certain things we can't use this variable for which is a trap for the unwary).  Maybe 
```
command: 'echo {{ output_dir }}'
```
or something else like that would be a better way to set this.

pilou- replied:
Updated using `command` task (https://github.com/ansible/ansible/pull/25672/commits/942075ef14e97a502ed023839d73bc66fe0cc2f3).

abadger commented on test/integration/targets/copy/tasks/main.yml:
```diff
@@ -174,7 +174,7 @@
   copy:
-    src: '{{ output_file }}'
+    src: 'foo.txt'
     dest: '{{ output_dir }}/copy-foo.txt'
```
This is incorrect.  We have to transfer from ``` { output_file }}``` becaue we've set the mode of that file to 0547.

pilou- replied:
Fixed by using `connection: local` in another commit.

abadger commented on test/integration/targets/copy/tasks/main.yml:
```diff
@@ -169,8 +169,9 @@
 - name: Set file perms to an odd value
   file:
-    path: '{{ output_file }}'
+    path: '{{ role_path }}/files/foo.txt'
```
I do not like doing this because then it's changing the mode of a file that is checked into the git repo.

abadger replied:
workaround: copy the file to another directory on the controller side and then do the copy to the remote machine from there.

pilou- replied:
Updated: another `foo.txt` file is created on the controller side, in the temporary directory (https://github.com/ansible/ansible/pull/25672/commits/e0a71a51276a528c2c013c6e4cde8d36d844ec6b#diff-6ce41036c58266d9f14a165103092878).

abadger commented on test/integration/targets/copy/tasks/main.yml:
```diff
@@ -588,36 +590,38 @@
+- block:
+  - name: Create a directory to copy from
+    file:
+      path: '{{ role_path }}/tmp/source1'
```
I don't think role_path is an appropriate location to be creating more files.  Doesn't a later commit switch us to using a different user?  Maybe tmpdir (/tmp/ or /var/tmp/) is the only choice that makes sense here?

abadger replied:
ah, not on the controller... I think tmpdir is still a better choice, though.

pilou- replied:
Updated: a temporary directory is used on the controller side (https://github.com/ansible/ansible/pull/25672/commits/e0a71a51276a528c2c013c6e4cde8d36d844ec6b#diff-cb51fd8205ec7b17fb119bd1a123ed2d).

abadger commented:
Okay, four areas noted (I think that one was addressed in a subsequent commit so really only three types of things that need to be changed).

pilou- committed:
Checks must not expect output of tested module to be right

pilou- committed:
Use a temporary directory on the controller

pilou- committed:
Use sha1 & md5 filters instead of hardcoded values

pilou- committed:
Use 'remote_dir' for directory on managed host

abadger commented:
Looks good.  +1 to merge once shippable passes.

pilou- commented:
Updated as requested.
Added:
- `remote_dir` is used instead of `output_dir`
- `sha1` and `md5` filters are used instead of hardcoded values (it's easier to debug when something goes wrong)

abadger commented:
Ugh.  Looks like your copy test is exposing a bug in the tempfile module on
OSX.  It looks like a bad interaction between OSX's python and the naming
of the AnsiBallZ file we ship over the wire.  I'll have to look into that.
In the meantime, can you use something other than the tempfile module?

On Mon, Jul 31, 2017 at 7:18 PM, Pilou <notifications@github.com> wrote:

> Updated as requested.
> Added:
>
>    - remote_dir is used instead of output_dir
>    - sha1 and md5 filters are used instead of hardcoded values (it's
>    easier to debug when something goes wrong)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ansible/ansible/pull/25672#issuecomment-319248707>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMxWjkzYhWJWKkE57PKW8nKnwf1OPvGks5sTorigaJpZM4N5Op_>
> .
>

pilou- committed:
Workaround tempfile error on OS X

Error was:
temp_path = tempfile.mkdtemp(prefix='ansible_')
AttributeError: 'module' object has no attribute 'mkdtemp'"

pilou- commented:
Workaround added, `tempfile` problem looks like https://bugs.launchpad.net/fuel/+bug/1595156.

abadger commented:
Thanks @pilou- merged to devel for the 2.4 release.
